driver-sync/src/test/functional/com/mongodb/client/unified/UnifiedTest.java

```diff
@@ -0,0 +1,384 @@
+        if (runOnRequirements != null) {
+            assumeTrue("Run-on requirements not met", runOnRequirementsMet(runOnRequirements, getServerVersion()));
+        }
+        if (definition.containsKey("runOnRequirements")) {
+            assumeTrue("Run-on requirements not met",
+                    runOnRequirementsMet(definition.getArray("runOnRequirements", new BsonArray()), getServerVersion()));
+        }
+        if (definition.containsKey("skipReason")) {
+            throw new AssumptionViolatedException(definition.getString("skipReason").getValue());
+        }
+        entities.init(entitiesArray, this::createMongoClient);
+        addInitialData();
+    }
+
+    @After
+    public void cleanUp() {
+        for (FailPoint failPoint : failPoints) {
+            failPoint.disableFailPoint();
+        }
+        entities.close();
+    }
+
+    @Test
+    public void shouldPassAllOutcomes() {
+        for (BsonValue cur : definition.getArray("operations")) {
+            assertOperation(cur.asDocument());
+        }
+
+        if (definition.containsKey("outcome")) {
+            assertOutcome();
+        }
+
+        if (definition.containsKey("expectEvents")) {
+            compareEvents(definition);
+        }
+    }
+
+    private void compareEvents(final BsonDocument operation) {
+        for (BsonValue cur : operation.getArray("expectEvents")) {
+            BsonDocument curClientEvents = cur.asDocument();
+            TestCommandListener listener = entities.getClientCommandListener(curClientEvents.getString("client").getValue());
+            eventMatcher.assertEventsEquality(curClientEvents.getArray("events"), listener.getEvents());
+        }
+    }
+
+    private void assertOutcome() {
+        for (BsonValue cur : definition.getArray("outcome")) {
+            BsonDocument curDocument = cur.asDocument();
+            MongoNamespace namespace = new MongoNamespace(curDocument.getString("databaseName").getValue(),
+                    curDocument.getString("collectionName").getValue());
+            List<BsonDocument> collectionData = new CollectionHelper<>(new BsonDocumentCodec(), namespace).find();
+            assertEquals(curDocument.getArray("documents").stream().map(BsonValue::asDocument).collect(toList()),
+                    collectionData);
+        }
+    }
+
+    private void assertOperation(final BsonDocument operation) {
+        OperationResult result = executeOperation(operation);
+        if (operation.containsKey("expectResult")) {
+            valueMatcher.assertValuesMatch(operation.get("expectResult"), requireNonNull(result.getResult()));
+        } else if (operation.containsKey("expectError")) {
+            ErrorMatcher.assertErrorsMatch(operation.getDocument("expectError"), requireNonNull(result.getException()));
+        }
+    }
+
+    private OperationResult executeOperation(final BsonDocument operation) {
+        String name = operation.getString("name").getValue();
+        switch (name) {
```

Comment:
I like this much more than the older reflection on the method name.  Easier to debug!

Reply:
Agreed!